### PR TITLE
Remove global use for call convention implementation

### PR DESCRIPTION
**Pull request by iotaaxel**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

- Added a reminder to remove the global variable(s).

### CODE REVIEW

**Comment by iotaaxel** [comment]
@trufae

**Comment by iotaaxel** [comment]
Ready for review. Can rebase again.

File: libr/anal/cc.c [hunk 1]
```diff
@@ -211,6 +211,30 @@ R_API void r_anal_cc_set_error(RAnal *anal, const char *convention, const char *
...
 }
 
+#if R2_580
+R_API int r_anal_cc_max_arg(RAnal *anal, const char *cc, void *oldDB, char *oldCC, int *oldArg) {
+	int i = 0;
+	r_return_val_if_fail (anal && DB && cc, 0);
+	*oldDB = NULL;
+	*oldCC = NULL;
+	*oldArg = 0;
+	if (oldDB == DB && !strcmp (cc, oldCC)) {
```
Comment (trufae) [inline libr/anal/cc.c:221]:
```suggestion
	if (max_cc_cached (anal, cc, DB)) {
```

File: libr/anal/cc.c [hunk 2]
```diff
@@ -211,6 +211,30 @@ R_API void r_anal_cc_set_error(RAnal *anal, const char *convention, const char *
 	r_strbuf_fini (&sb);
 }
 
+#if R2_580
```
Comment (trufae) [inline libr/anal/cc.c:214]: typedef struct {
	void *db;
	 char *cc;
	 int oldArg;
} RAnalCCMaxCache;

Comment (trufae) [reply libr/anal/cc.c:214]: store this into anal->ccmax_cache

File: libr/anal/cc.c [hunk 3]
```diff
@@ -211,6 +211,30 @@ R_API void r_anal_cc_set_error(RAnal *anal, const char *convention, const char *
...
+R_API int r_anal_cc_max_arg(RAnal *anal, const char *cc, void *oldDB, char *oldCC, int *oldArg) {
+	int i = 0;
+	r_return_val_if_fail (anal && DB && cc, 0);
```
Comment (trufae) [inline libr/anal/cc.c:217]:
```suggestion
	r_return_val_if_fail (anal && cc, 0);
```

File: libr/anal/cc.c [hunk 4]
```diff
@@ -211,6 +211,30 @@ R_API void r_anal_cc_set_error(RAnal *anal, const char *convention, const char *
 	r_strbuf_fini (&sb);
 }
 
+#if R2_580
+R_API int r_anal_cc_max_arg(RAnal *anal, const char *cc, void *oldDB, char *oldCC, int *oldArg) {
```
Comment (trufae) [inline libr/anal/cc.c:215]:
```suggestion
R_API int r_anal_cc_max_arg(RAnal *anal, const char *cc) {
```

File: libr/anal/cc.c (hunk 2 above)
Comment (trufae) [reply libr/anal/cc.c:214]:
```
static bool max_cc_cached (RAnal *anal, const char *cc, void *db) {
  return (anal->ccmax_cache.db == db && anal->ccmax_cache.oldCC && anal->ccmax_cache.oldCC.cc == cc);
}
```

File: libr/anal/cc.c [hunk 5]
```diff
@@ -211,6 +211,30 @@ R_API void r_anal_cc_set_error(RAnal *anal, const char *convention, const char *
...
+			break;
+		}
+	}
+	oldArg = i;
+	return i;
+}
+#else
 R_API int r_anal_cc_max_arg(RAnal *anal, const char *cc) {
 	int i = 0;
 	r_return_val_if_fail (anal && DB && cc, 0);
```
Comment (trufae) [inline libr/anal/cc.c:240]: remove the R_TH_LOCALs below. they can be cached inside the RAnal->ccmax_cache struct

File: libr/anal/cc.c [hunk 6]
```diff
@@ -211,6 +211,30 @@ R_API void r_anal_cc_set_error(RAnal *anal, const char *convention, const char *
...
+R_API int r_anal_cc_max_arg(RAnal *anal, const char *cc, void *oldDB, char *oldCC, int *oldArg) {
+	int i = 0;
+	r_return_val_if_fail (anal && DB && cc, 0);
+	*oldDB = NULL;
+	*oldCC = NULL;
+	*oldArg = 0;
+	if (oldDB == DB && !strcmp (cc, oldCC)) {
+		return oldArg;
+	}
+	oldDB = DB;
```
Comment (trufae) [inline libr/anal/cc.c:224]: anal->ccmax_cache.db = DB;
free (anal->ccmax_cache.cc );
anal->ccmax_cache.cc = strdup(cc);

Comment (trufae) [reply libr/anal/cc.c:224]: in r_anal_free () { you need to free the anal->ccmax_cache.cc)

File: libr/core/anal_tp.c [hunk 7]
```diff
@@ -312,7 +312,14 @@ static void type_match(RCore *core, char *fcn_name, ut64 addr, ut64 baddr, const
 	}
 	if (!max) {
 		if (!in_stack) {
+#if R2_580
```
Comment (trufae) [inline libr/core/anal_tp.c:315]: remove that #if block

File: libr/core/canal.c [hunk 8]
```diff
@@ -3526,7 +3526,14 @@ static bool anal_block_cb(RAnalBlock *bb, BlockRecurseCtx *ctx) {
 		}
 		if (optype == R_ANAL_OP_TYPE_CALL) {
 			int i;
+#if R2_580
```
Comment (trufae) [inline libr/core/canal.c:3529]: same here

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/anal/cc.c (hunk 2 above)
Comment (iotaaxel) [reply libr/anal/cc.c]: Keeping the structs similar as https://github.com/radareorg/radare2/pull/20443

**Pull request closed without merging**